File: 2015/03/nim/solution.nim
import options
import sets
import streams

# Direction --------------------------------------------------------------------

type Direction = enum
  north, south, east, west

proc parse(direction: char): Option[Direction] =
  case direction
  of '^':
    some(north)
  of 'v':
    some(south)
  of '>':
    some(east)
  of '<':
    some(west)
  else:
    none(Direction)


# Point ------------------------------------------------------------------------

type Point = tuple[x, y: int]

let originPoint: Point =
  (0, 0)

proc move(point: var Point, direction: Direction) =
  case direction
  of north:
    point.y += 1
  of south:
    point.y -= 1
  of east:
    point.x += 1
  of west:
    point.x -= 1


# Tracker ----------------------------------------------------------------------

type Tracker = ref object of RootObj
  houses: ref HashSet[Point]

proc newTracker: Tracker =
  let houses = new HashSet[Point]
  houses[].init()
  Tracker(houses: houses)

proc recordVisit(t: Tracker, p: Point) =
  t.houses[].incl(p)

proc housesVisited(t: Tracker): int =
  t.houses[].len


# Santa ------------------------------------------------------------------------

type Santa = ref object of RootObj
  position: Point
  tracker: Tracker

proc move(s: var Santa, dir: Direction) =
  s.position.move(dir)
  s.tracker.recordVisit(s.position)

proc newSanta(t: Tracker): Santa =
  result = Santa(position: originPoint, tracker: t)
  t.recordVisit(result.position)


# Main -------------------------------------------------------------------------

iterator directions: Direction {.inline.} =
  var
    direction: Option[Direction]
    input = newFileStream(stdin)
  while not input.atEnd:
    direction = input.readChar().parse()
    if direction.isSome:
      yield direction.get

proc main =
  var
    loneTracker = newTracker()
    loneSanta = newSanta(loneTracker)
    pairTracker = newTracker()
    pairSanta = newSanta(pairTracker)
    roboSanta = newSanta(pairTracker)
    currentSanta = pairSanta

  proc swap(s: var Santa) =
    if s == pairSanta:
      s = roboSanta
    else:
      s = pairSanta

  for direction in directions():
    loneSanta.move(direction)
    currentSanta.move(direction)
    swap(currentSanta)

  echo "Houses Visited (Part One): " & $loneTracker.housesVisited
  echo "Houses Visited (Part Two): " & $pairTracker.housesVisited

main()

File: 2015/03/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ALLOWED_VISITED_HOUSES 5000

typedef struct house house;
struct house
  {
    int x;
    int y;
  };

void
move (house *current, char direction)
{
  switch (direction)
    {
      case '>':
        current->x++;
        break;
      case '<':
        current->x--;
        break;
      case '^':
        current->y++;
        break;
      case 'v':
        current->y--;
        break;
    }
}

typedef struct visit_log visit_log;
struct visit_log
  {
    house log[MAX_ALLOWED_VISITED_HOUSES];
    int count;
  };

void
record_visit (visit_log *log, house current)
{
  for (int i = 0; i < log->count; i++)
    {
      house past = log->log[i];
      if (past.x == current.x && past.y == current.y)
        return;
    }

  if ((log->count + 1) == MAX_ALLOWED_VISITED_HOUSES)
    {
      fprintf(stderr, "ERROR: Visited too many houses\n");
      exit(1);
    }
  
  log->log[log->count] = current;
  log->count += 1;
}

void
swap (house **current, house *santa, house *robot)
{
  if (*current == santa)
    *current = robot;
  else
    *current = santa;
}

int
main (int argc, char **argv)
{
  visit_log lone_log;
  lone_log.count = 0;

  house lone_santa = { .x = 0, .y = 0 };
  record_visit(&lone_log, lone_santa);

  visit_log pair_log;
  pair_log.count = 0;

  house pair_santa = { .x = 0, .y = 0 };
  house robo_santa = { .x = 0, .y = 0 };
  house *current_santa = &pair_santa;
  record_visit(&pair_log, *current_santa);

  char command;
  while ((command = fgetc(stdin)) != EOF)
    {
      // Part 1
      move(&lone_santa, command);
      record_visit(&lone_log, lone_santa);

      // Part 2
      move(current_santa, command);
      record_visit(&pair_log, *current_santa);
      swap(&current_santa, &pair_santa, &robo_santa);
    }

  printf("Part 1: %d\n", lone_log.count);
  printf("Part 2: %d\n", pair_log.count);
}
